pyomo/core/util.py

```diff
@@ -76,14 +76,18 @@ def quicksum(args, start=0, linear=None):
             #
             # Get the first term, which we will test for linearity
             #
-            try:
-                first = next(args, None)
-            except:
-                try:
+            if '__next__' not in args.__dir__():
+                # Since `IndexedComponent_slice` implements `__getattr__`,
+                # `hasattr(args,'__next__')` can not distinguish
+                # `IndexedComponent_slice` from `_IndexedComponent_slice_iter'
+                if hasattr(args, '__iter__'):
                     args = args.__iter__()
-                    first = next(args, None)
-                except:
+                else:
                     raise RuntimeError("The argument to quicksum() is not iterable!")
+            # No `try...except...`
+            # to prevent hiding exceptions raised in the iterator from the user
+            first = next(args, None)
```

Comment:
Wouldn't it be simpler to just do the following:
```python
# Ensure that args is an iterator (this manages things like IndexedComponent_slice objects)
args = iter(args)
first = next(args, None)
```

Reply:
Thank you for the suggestion!

- I have put this line of code to the beginning of the `quicksum` function since `args` is required to be iterable for all conditions.
- 2 test for `quicksum` was added to `TestSumExpression` in `test_numeric_expr.py`.